ocaml/xapi/xapi_vm_snapshot.ml

```diff
@@ -478,6 +479,36 @@ let revert_vm_fields ~__context ~snapshot ~vm =
     ~overrides ;
   TaskHelper.set_progress ~__context 0.1
 
+(* See CP-40528. A snapshot is a VM record that contains the record of
+   the VM at snapshot time in snapshot_metadata, including its VTPM.
+   Most data is restored from snapshot_metadata. At the time the
+   snapshot is taken, a copy of the VTPM is taken and this is part of
+   the VM record. We want to use a copy of this VTPM. We need to make a
+   copy because a snapshot can be restored multiple times and we have to
+   ensure the VTPM of the snapshot remains unaltered. *)
+
+let update_vtpm ~__context ~snapshot ~vm =
+  debug "%s for VM %s (1/2)" __FUNCTION__ (Ref.string_of vm) ;
+  let vtpms = Db.VM.get_VTPMs ~__context ~self:vm in
+  let uuids =
+    vtpms |> List.map (fun self -> Db.VTPM.get_uuid ~__context ~self)
+  in
+  (* destroy current VTPM *)
+  List.iter (fun vtpm -> Xapi_vtpm.destroy ~__context ~self:vtpm) vtpms ;
```

Comment:
The function `update_vifs_vbds_vgpus_and_vusbs` above uses `safe_destroy_X` functions that don't raise an error if the ref no longer exists for some reason. I suppose this is to make reverting to a snapshot robust if the current VM is corrupted for some reason.